neqo-transport/src/ackrate.rs

```diff
@@ -0,0 +1,217 @@
+    packets: usize,
+    /// The maximum delay before sending an ACK.
+    delay: Duration,
+}
+
+impl AckRate {
+    pub fn new(minimum: Duration, ratio: u8, cwnd: usize, mtu: usize, rtt: Duration) -> Self {
+        const PACKET_RATIO: usize = ACK_RATIO_SCALE as usize;
+        // At worst, ask for an ACK for every other packet.
+        const MIN_PACKETS: usize = 2;
+        // At worst, require an ACK every 256 packets.
+        const MAX_PACKETS: usize = 256;
+        const RTT_RATIO: u32 = ACK_RATIO_SCALE as u32;
+        const MAX_DELAY: Duration = Duration::from_millis(50);
+
+        let packets = cwnd * PACKET_RATIO / mtu / usize::from(ratio);
+        let packets = max(MIN_PACKETS, min(packets, MAX_PACKETS)) - 1;
+        let delay = rtt * RTT_RATIO / u32::from(ratio);
+        let delay = max(minimum, min(delay, MAX_DELAY));
+        qtrace!(
+            "Set ACK rate: {}/{}/{}, {:?} = {}, {:?}",
+            cwnd,
+            mtu,
+            ratio,
+            rtt,
+            packets,
+            delay,
+        );
+        Self { packets, delay }
+    }
+
+    pub fn write_frame(&self, builder: &mut PacketBuilder, seqno: u64) -> bool {
+        write_varint_frame(
+            builder,
+            &[
+                FRAME_TYPE_ACK_FREQUENCY,
+                seqno,
+                u64::try_from(self.packets + 1).unwrap(),
+                u64::try_from(self.delay.as_micros()).unwrap(),
+                0,
+            ],
+        )
+    }
+
+    /// Determine whether to send an update frame.
+    pub fn needs_update(&self, other: &Self) -> bool {
+        // Allow more flexibility for delays, as those can change
+        // by small amounts fairly easily.
+        self.packets != other.packets
+            || other.delay * 2 < self.delay
+            || other.delay > self.delay * 2
+    }
+}
+
+#[derive(Debug)]
+pub struct FlexibleAckRate {
+    current: AckRate,
+    target: AckRate,
+    next_frame_seqno: u64,
+    frame_outstanding: bool,
+    min_ack_delay: Duration,
+    ratio: u8,
+}
+
+impl FlexibleAckRate {
+    fn new(
+        max_ack_delay: Duration,
+        min_ack_delay: Duration,
+        ratio: u8,
+        cwnd: usize,
+        mtu: usize,
+        rtt: Duration,
+    ) -> Self {
+        qtrace!("ACK rate: {:?} {:?}", max_ack_delay, min_ack_delay);
+        let ratio = max(ACK_RATIO_SCALE, ratio); // clamp it
+        Self {
+            current: AckRate {
+                packets: 1,
+                delay: max_ack_delay,
+            },
+            target: AckRate::new(min_ack_delay, ratio, cwnd, mtu, rtt),
+            next_frame_seqno: 0,
+            frame_outstanding: false,
+            min_ack_delay,
+            ratio,
+        }
+    }
+
+    fn write_frames(
+        &mut self,
+        builder: &mut PacketBuilder,
+        tokens: &mut Vec<RecoveryToken>,
+        stats: &mut FrameStats,
+    ) {
+        if !self.frame_outstanding
```

Comment:
Right.  I think that any time we change our mind and we have an outstanding packet, we just change to writing the new one rather than a repeat of the old one.  We are only updating a single piece of state.